man/ipeextract.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
.TH IPEEXTRACT 1 "June 27, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
ipeextract \- Extract Ipe XML stream from Ipe PDF or EPS files
.SH SYNOPSIS
.B ipeextract
\fIinput-file\fP [ \fIoutput-file\fP ]

.SH DESCRIPTION
.PP
\fBipeextract\fP extracts the Ipe XML stream from an Ipe document in
PDF or Postscript format generated by any version of Ipe 6 or Ipe 7.

If the \fIoutput-file\fP is not given, \fBipeextract\fP replaces the
input file extension by \fIxml\fP.


.SH ENVIRONMENT VARIABLES

\fBipeextract\fP respects the following environment variables:

.TP
\fBIPEDEBUG\fP
set this to 1 for debugging output from \fBipeextract\fP.

.SH AUTHOR
Otfried Cheong

.SH REPORTING BUGS
.ad l
Please report bugs using Ipe bugzilla at
.I "http://ipe7.sourceforge.net/bugzilla.html"

.SH SEE ALSO
.ad l
Ipe is documented fully in
.IR "The Ipe Manual" ,
which is available online at
.I "http://ipe7.sourceforge.net/manual/manual.html" .


news.txt
--------------------------------------------------------------------

Ipe 7.0.4 (2009-10-14) SVN 335

 * Fixed bug in save_as (bug #305).

 * ipe6upgrade: fixed empty dash.
 
 * Fixed handling of matrix in path join.

 * Added "arc" and "farc" arrow types.  When the final segment of a
   path is circular arc, then these arrows take on a shape that
   depends on the radius of the arc, as metafont arrows do.

 * Fixed handling of path objects with missing stroke and fill
   attributes. 

 * Added "presentation.isy" to styles directory.  You can now create a
   new presentation by saying "ipe -sheet presentation".

 * Fixed computation of bounding box of transformed text object.

 * Fixed handling of horizontal and vertical alignment and
   transformable text from Properties menu.

--------------------------------------------------------------------

Ipe 7.0.3 (2009-10-10) SVN 318

 * Style sheets now looked for in IPESTYLES directory, so it suffices
   to say "ipe -sheet presentation" to create a new presentations.

 * Fixed cropped boundaries in iperender (bug #284).

 * "Save as" dialog will start in more sensible directory (can be
   customized, see prefs.save_as_directory).

 * Using Cairo font if Latex has not been run.

 * Fixed colorstack problem in Latex run, require Pdflatex 1.30.

 * Ipe can now create independent bounding boxes for each view.
   This is activated by creating a layer named "VIEWBBOX".

--------------------------------------------------------------------

Ipe 7.0.2 (2009-09-27) SVN 311

 * Path edit tool, including scissors and path join implemented.

 * Implemented clipping path operations for groups.

 * Several bugs fixed.

 * Implemented automatic angular snapping.

 * Included intersection snapping (from pre29).

 * Fixed Cairo glyph cache handling (bug #293).

 * Fixed snapping to ends of B-spline (bug #290).

--------------------------------------------------------------------

Ipe 7.0.1 (2009-09-17) SVN 288

 * Path entry tool finished, can switch between spline and line mode
   and handle quadratic and cubic Bezier splines and circular arcs.

 * If Ipe is called with a non-existing file as an argument, it will
   not show an error message but remember the file for the next save.

 * Mark shape selection moved to properties window.

 * Added $(INSTALL_ROOT) variable to Makefiles for packaging.
 
--------------------------------------------------------------------

Ipe 7.0.0 (2009-09-16) SVN 282
 
 * Most of Ipe was rewritten entirely.

 * Group object can now have a clipping path.

 * User-defined marks and arrows can be defined in a style sheet.

 * Gradients, tilings, and transparency added.

 * Standard style settings are no longer hardwired into Ipe. 
   Everything is defined in loadable stylesheets.

 * Mouse button shortcuts have been changed.  Right mouse button is
   now always the popup menu. 

 * Using Cairo library for rendering Ipe documents, Qt for the user
   interface, and Lua as a scripting language.  The entire high-level
   Ipe model is written in Lua.

 * Namespaces ipe, ipeqt, ipeui, ipelua.

 * C++ method names now start with a lowercase letter.

 * Mark object removed (it is now implemented using symbols).

 * Ipeextract can open any Ipe file (any version of Ipe 6 or 7) and
   extract the XML stream. Ipeextract can help to recover files that
   are unreadable due to an Ipe bug.

 * There were many small changes in the file format during the
   lifetime of Ipe 6, and it was becoming cumbersome (and buggy) to
   keep Ipe able to read all these various versions.  Ipe 7.0 has a
   fresh format, and a separate program ipe6upgrade is used to convert
   from Ipe 6 to Ipe 7 format.

 * The grid has been replaced by graph paper.

 * Ipetopng was replaced by iperender, which can produce output in PNG
   and in SVG.  SVG output can be used to embed scalable Ipe figures
   in webpages.  Iperender does not need to be linked with a GUI
   toolkit and requires no X server, so it can easily be used on a
   webserver (to convert Latex source to images or SVG art).

 * Ipe now fails gracefully on incorrect XML input.

 * Painter now has a drawArc primitive.  Circular arcs on the screen
   should now be accurate even at high zoom. (PDF and Postscript
   output still needs to approximate using splines, though.)

 * Text "transformable" attribute replaced by general
   "transformations" attribute. Any object can be restricted to
   translations only or rigid motions only.

 * Ipe now runs creates temporary directory for running latex under
   Windows in user's local APPDATA directory.

 * Ipe XML format contains DOCTYPE declaration, and "ipe.dtd" is
   provided. 

 * Ipetoipe, ipe6upgrade, ipeextract guess output filename if it is
   not provided.

 * Appearance of page numbers on the page can be customized in the
   style sheet.

 * Text for latex conversion is now written out in UTF-8, and it is
   assumed that Pdflatex can handle that (using a suitable package,
   either "inputenc" or "CJK").

 * Ipe symbols can be used as the "bullet" in itemize environments etc.

 * Path objects can be drawn using a pen on a tablet PC.

 * New alignment modes to distribute centers and sides equally.

 * Latex and XML code can now be edited using an external editor.

 * Support for subpixel information from a tablet (but still
   completely untested).

--------------------------------------------------------------------

Ipe 6.0 preview 32 patch 1 (2009-04-30)

 * Fixed writing floating point numbers again (bug #279).

 * Qt 4.5 breaks setting locale (bug #281).

 * Fixed broken \rule (bug #268).

 * Removed check against freetype version.  It breaks Ipe whenever
   distributions upgrade freetype, and seems no longer necessary.  In
   principle, freetype should provide upwards binary compatibility.

--------------------------------------------------------------------

Ipe 6.0 preview 32 (2008-12-01)

 * Fixed writing floating point numbers (bug #274).


src/ipeextract/ipeextract.cpp
// --------------------------------------------------------------------
// ipeextract
// --------------------------------------------------------------------
/*

    This file is part of the extensible drawing editor Ipe.
    Copyright (C) 1993-2009  Otfried Cheong

    Ipe is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    As a special exception, you have permission to link Ipe with the
    CGAL library and distribute executables, as long as you follow the
    requirements of the Gnu General Public License in regard to all of
    the software in the executable aside from CGAL.

    Ipe is distributed in the hope that it will be useful, but WITHOUT
    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
    or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
    License for more details.

    You should have received a copy of the GNU General Public License
    along with Ipe; if not, you can find it at
    "http://www.gnu.org/copyleft/gpl.html", or write to the Free
    Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

*/

#include "ipexml.h"
#include "ipeutils.h"
#include "ipepdfparser.h"

using namespace ipe;

// ---------------------------------------------------------------------

enum TFormat {EXml, EPdf, EEps, EIpe5, EUnknown};

String readLine(DataSource &source)
{
  String s;
  int ch = source.getChar();
  while (ch != EOF && ch != '\n') {
    s += char(ch);
    ch = source.getChar();
  }
  return s;
}

//! Determine format of file in \a source.
TFormat fileFormat(DataSource &source)
{
  String s1 = readLine(source);
  String s2 = readLine(source);
  if (s1.substr(0, 5) == "<?xml" || s1.substr(0, 4) == "<ipe")
    return EXml;
  if (s1.substr(0, 4) == "%PDF")
    return EPdf;  // let's assume it contains an Ipe stream
  if (s1.substr(0, 4) == "%!PS") {
    if (s2.substr(0, 11) != "%%Creator: ")
      return EUnknown;
    if (s2.substr(11, 6) == "Ipelib" || s2.substr(11, 4) == "xpdf")
      return EEps;
    if (s2.substr(11, 3) == "Ipe")
      return EIpe5;
    return EUnknown;
  }
  if (s1.substr(0, 5) == "%\\Ipe" || s1.substr(0, 6) == "%\\MIPE")
    return EIpe5;
  return EUnknown;
}

// --------------------------------------------------------------------

class StreamParser : public XmlParser {
public:
  explicit StreamParser(DataSource &source, std::FILE *out)
    : XmlParser(source), iOut(out) { /* nothing */ }
  bool parse();
  virtual Buffer image(int objNum) = 0;
  void writeAttributes(const XmlAttributes &attr);
  bool parseBitmap();
private:
  std::FILE *iOut;
};

bool StreamParser::parse()
{
  while (!eos()) {
    bool lt = (iCh == '<');
    fputc(iCh, iOut);
    getChar();
    // look out for <bitmap> tag
    if (lt && iCh == 'b') {
      String tag;
      while (isTagChar(iCh)) {
	tag += char(iCh);
	fputc(iCh, iOut);
	getChar();
      }
      // at char after tag
      if (tag == "bitmap" && !parseBitmap())
	return false;
    }
  }
  return true;
}

// write out attributes, but drop 'pdfObject'
void StreamParser::writeAttributes(const XmlAttributes &attr)
{
  for (XmlAttributes::const_iterator it = attr.begin();
       it != attr.end(); ++it)
    if (it->first != "pdfObject")
      fprintf(iOut, " %s=\"%s\"", it->first.z(), it->second.z());
  fprintf(iOut, ">\n");
}

bool StreamParser::parseBitmap()
{
  XmlAttributes attr;
  if (!parseAttributes(attr))
    return false;
  String objNumStr;
  if (attr.slash() && attr.has("pdfObject", objNumStr)) {
    int objNum = Lex(objNumStr).getInt();
    Buffer bits = image(objNum);
    writeAttributes(attr);
    // write bits
    const char *data = bits.data();
    const char *fin = data + bits.size();
    int col = 0;
    while (data != fin) {
      fprintf(iOut, "%02x", (*data++ & 0xff));
      if (++col == 36) {
	fputc('\n', iOut);
	col = 0;
      }
    }
    if (col > 0)
      fputc('\n', iOut);
    fprintf(iOut, "</bitmap>\n");
  } else {
    // just write out attributes
    writeAttributes(attr);
  }
  return true;
}

// --------------------------------------------------------------------

class StreamParserPdf : public StreamParser {
public:
  explicit StreamParserPdf(PdfFile &loader, DataSource &source,
			   std::FILE *out)
    : StreamParser(source, out), iLoader(loader) { /* nothing */ }
  virtual Buffer image(int objNum);
private:
  PdfFile &iLoader;
};

Buffer StreamParserPdf::image(int objNum)
{
  const PdfObj *obj = iLoader.object(objNum);
  if (!obj || !obj->dict() || obj->dict()->stream().size() == 0)
    return Buffer();
  return obj->dict()->stream();
}

// --------------------------------------------------------------------

class PsSource : public DataSource {
public:
  PsSource(DataSource &source) : iSource(source) { /* nothing */ }
  bool skipToXml();
  String readLine();
  Buffer image(int index) const;
  int getNext() const;
  inline bool deflated() const { return iDeflated; }

  virtual int getChar();
private:
  DataSource &iSource;
  std::vector<Buffer> iImages;
  bool iEos;
  bool iDeflated;
};

int PsSource::getChar()
{
  int ch = iSource.getChar();
  if (ch == '\n')
    iSource.getChar(); // remove '%'
  return ch;
}

String PsSource::readLine()
{
  String s;
  int ch = iSource.getChar();
  while (ch != EOF && ch != '\n') {
    s += char(ch);
    ch = iSource.getChar();
  }
  iEos = (ch == EOF);
  return s;
}

Buffer PsSource::image(int index) const
{
  if (1 <= index && index <= int(iImages.size()))
    return iImages[index - 1];
  else
    return Buffer();
}

bool PsSource::skipToXml()
{
  iDeflated = false;

  String s1 = readLine();
  String s2 = readLine();

  if (s1.substr(0, 11) != "%!PS-Adobe-" ||
      s2.substr(0, 11) != "%%Creator: ")
    return false;

  if (s2.substr(11, 6) == "Ipelib") {
    // the 'modern' file format of Ipe 6.0 preview 17 and later
    do {
      s1 = readLine();
      if (s1.substr(0, 17) == "%%BeginIpeImage: ") {
	Lex lex(s1.substr(17));
	int num, len;
	lex >> num >> len;
	if (num != int(iImages.size() + 1))
	  return false;
	(void) readLine();  // skip 'image'
	Buffer buf(len);
	A85Source a85(iSource);
	char *p = buf.data();
	char *p1 = p + buf.size();
	while (p < p1) {
	  int ch = a85.getChar();
	  if (ch == EOF)
	    return false;
	  *p++ = char(ch);
	}
	iImages.push_back(buf);
      }
    } while (!iEos && s1.substr(0, 13) != "%%BeginIpeXml");

    iDeflated = (s1.substr(13, 14) == ": /FlateDecode");

  } else {
    // the 'old' file format generated through pdftops
    do {
      s1 = readLine();
    } while (!iEos && s1.substr(0, 10) != "%%EndSetup");
  }
  if (iEos)
    return false;
  (void) iSource.getChar(); // skip '%' before <ipe>
  return true;
}

// --------------------------------------------------------------------

class StreamParserPs : public StreamParser {
public:
  explicit StreamParserPs(PsSource &loader, DataSource &source,
			  std::FILE *out)
    : StreamParser(source, out), iLoader(loader) { /* nothing */ }
  virtual Buffer image(int objNum);
private:
  PsSource &iLoader;
};

Buffer StreamParserPs::image(int objNum)
{
  return iLoader.image(objNum);
}

// --------------------------------------------------------------------

static bool extractPs(DataSource &source, std::FILE *out)
{
  PsSource psSource(source);
  if (!psSource.skipToXml()) {
    fprintf(stderr, "Could not find XML stream.\n");
    return false;
  }

  if (psSource.deflated()) {
    A85Source a85(psSource);
    InflateSource source(a85);
    StreamParserPs parser(psSource, source, out);
    return parser.parse();
  } else {
    StreamParserPs parser(psSource, psSource, out);
    return parser.parse();
  }
  return false;
}

static bool extractPdf(DataSource &source, std::FILE *out)
{
  PdfFile loader;
  if (!loader.parse(source)) {
    fprintf(stderr, "Error parsing PDF file\n");
    return false;
  }

  // try ancient format version first (early previews of Ipe 6.0)
  const PdfObj *obj = loader.catalog()->get("Ipe", &loader);
  if (!obj) {
    obj = loader.object(1);
    if (!obj || !obj->dict()) {
      fprintf(stderr, "Input file does not contain an Ipe XML stream.\n");
      return false;
    }

    const PdfObj *type = obj->dict()->get("Type", 0);
    if (!type || !type->name() || type->name()->value() != "Ipe") {
      fprintf(stderr, "Input file does not contain an Ipe XML stream.\n");
      return false;
    }
  }

  Buffer buffer = obj->dict()->stream();
  BufferSource xml(buffer);

  if (obj->dict()->deflated()) {
    InflateSource xml1(xml);
    StreamParserPdf parser(loader, xml1, out);
    return parser.parse();
  } else {
    StreamParserPdf parser(loader, xml, out);
    return parser.parse();
  }
}

// --------------------------------------------------------------------

static void usage()
{
  fprintf(stderr,
	  "Usage: ipeextract ( <input.pdf> | <input.eps> ) [<output.xml>]\n"
	  "Ipeextract extracts the XML stream from a PDF or Postscript file\n"
	  "generated by any version of Ipe 6 or Ipe 7.\n"
	  );
  exit(1);
}

int main(int argc, char *argv[])
{
  Platform::initLib(IPELIB_VERSION);

  // ensure one or two arguments
  if (argc != 2 && argc != 3)
    usage();

  const char *src = argv[1];
  String dst;

  if (argc == 3) {
    dst = argv[2];
  } else {
    String s = src;
    if (s.right(4) == ".pdf" || s.right(4) == ".eps")
      dst = s.left(s.size() - 3) + "xml";
    else
      dst = s + ".xml";
  }

  std::FILE *fd = std::fopen(src, "rb");
  if (!fd) {
    std::fprintf(stderr, "Could not open '%s'\n", src);
    exit(1);
  }
  FileSource source(fd);
  TFormat format = fileFormat(source);
  if (format == EXml) {
    fprintf(stderr, "Input file is already in XML format.\n");
  } else if (format == EIpe5) {
    fprintf(stderr, "Input file is in Ipe5 format.\n"
	    "Run 'ipe5toxml' to convert it to XML format.\n");
  } else {
    std::rewind(fd);
    std::FILE *out = std::fopen(dst.z(), "wb");
    if (!out) {
      fprintf(stderr, "Could not open '%s' for writing.\n", dst.z());
    } else {
      bool res = (format == EPdf) ?
	extractPdf(source, out) : extractPs(source, out);
      if (!res)
	fprintf(stderr, "Error during extraction of XML stream.\n");
      std::fclose(out);
    }
  }
  std::fclose(fd);
  return 0;
}

// --------------------------------------------------------------------
